Add coordinator wrapper

xianny opened the pull request:
## Description

Adds a wrapper for handling Coordinator transactions

## Testing instructions

`cd packages/contract-wrappers && yarn build && yarn test`

Unit tests cover fills, soft cancels, hard cancels, and handling server errors.

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

* New feature (non-breaking change which adds functionality)

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        feeRecipientAddress: string,
+    ): Promise<CoordinatorServerApprovalResponse | CoordinatorServerCancellationResponse> {
+        assert.doesConformToSchema('transaction', transaction, schemas.zeroExTransactionSchema);
+        assert.isETHAddressHex('feeRecipientAddress', feeRecipientAddress);
```
These assertions are not needed. Typescript does a good enough job with verifying these types at run-time. Remove.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+            throw new Error(
+                `Could not find endpoint for coordinator operator with { coordinatorAddress: ${
+                    this.address
+                }, registryAddress: ${this.registryAddress}, operatorAddress: ${feeRecipientAddress}}`,
```
This error message feels a bit confusing to me. How about:

```
throw new Error(`No Coordinator server endpoint found in Coordinator registry for feeRecipientAddress: ${feeRecipientAddress}. Registry contract address: ${this.registryAddress}`);
```

xianny replied:
this sounds way better, thanks!

dekz commented on packages/contract-wrappers/package.json:
```diff
@@ -39,6 +39,7 @@
     },
     "devDependencies": {
         "@0x/contracts-test-utils": "^3.1.2",
+        "@0x/coordinator-server": "^0.0.6",
```
It looks like this is bringing in some old dependencies of contract-wrappers, might result in some strange behaviour.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        assert.doesConformToSchema('transaction', transaction, schemas.zeroExTransactionSchema);
+        assert.isETHAddressHex('feeRecipientAddress', feeRecipientAddress);
+
+        const coordinatorOperatorEndpoint = await this._registryInstance.getCoordinatorEndpoint.callAsync(
```
I feel strongly we should do this step _before_ we ask the user to _sign_ the 0x transaction. If the endpoint doesn't exist in the registry, the signing process is a waste of time. Can we move this up?

xianny replied:
Yep, agree. I'll fix it

dekz commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        await assert.isSenderAddressAsync('makerAddress', makerAddress, this._web3Wrapper);
+
+        const data = this._transactionEncoder.batchCancelOrdersTx(orders);
+        const transaction = await this._toSignedZeroExTransactionAsync(data, makerAddress);
```
Do we need to do it this way for hard cancels? The resulting UX would be 2 MM prompts wouldn't it?

xianny replied:
Not sure I understand - is that because the Exchange method also does a MM prompt? 

Looks like the coordinator contract under the hood skips everything for cancels and forwards directly to the Exchange contract so we could call the exchange method directly.

dekz replied:
Ignore this my brain wasn't working!

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+
+        const requestPayload = {
+            signedTransaction: transaction,
+            txOrigin: senderAddress,
```
`txOrigin` should default  to `takerAddress`. `txOrigin` specifies who can ultimately submit the Ethereum transaction to the network.

xianny replied:
This is passed in from any calling methods but the variable is named `txOrigin` for consistency with the request payload schema that the server expects. When called for a fill approval it's passed as the takerAddress, and when called for a soft cancellation it's the makerAddress. Is that right?

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        );
+
+        if (response.status !== HTTP_OK) {
+            throw new Error(`${response.status}: ${JSON.stringify(await response.json())}`); // todo (xianny)
```
Is this error message going to improved still?

xianny replied:
yep!

xianny replied:
updated it, what do you think?

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        //     [],
+        //     [],
+        //     orderTransactionOpts,
+        // );
```
Is this outstanding?

xianny replied:
Yeah. Not sure why it doesnt work. It looks like the Coordinator contract should just be forwarding the transaction to the Exchange contract. Cancellations skip approval verification.

dekz replied:
It must go through the Coordinator correct? Otherwise the `cancelOrdersUpTo` will cancel orders for tuple
`(makerAddress, NULL_ADDRESS) with salt > targetOrderEpoch`. Rather than `(makerAddress, coordinatorAddress) with salt > targetOrderEpoch `

xianny replied:
Ahhh ok, I see what I was doing wrong. the test was retrieving orders for  (makerAddress, NULL_ADDRESS) instead of (makerAddress, coordinatorAddress). fixed it :)

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        return this._contractInstance.getSignerAddress.callAsync(hash, signature);
+    }
+
+    private async _sendTransactionAsync(
```
```suggestion
    private async _submitCoordinatorTransactionAsync(
```

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+    }
+
+    /**
+     * Fills a signed order with an amount denominated in baseUnits of the taker asset.
```
Let's put a more detailed description of what these methods do on all the subsequent methods as well.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        takerAddress: string,
+        orderTransactionOpts: OrderTransactionOpts = { shouldValidate: true },
+    ): Promise<string> {
+        assert.doesConformToSchema('signedOrders', signedOrders, schemas.signedOrdersSchema);
```
Add assertions for all the params of this method.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        for (const takerAssetFillAmount of takerAssetFillAmounts) {
+            assert.isBigNumber('takerAssetFillAmount', takerAssetFillAmount);
+        }
+        const feeRecipientAddress = getFeeRecipientOrThrow(signedOrders);
```
There can be more then one `feeRecipientAddress` in the supplied orders. If there are multiple coordinator server involved in a batch fill, the client must send a request to every coordinator server and must retrieve a signature from each. Sorry if this was not clear.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+     * @return Transaction hash.
+     */
+    @decorators.asyncZeroExErrorHandler
+    public async matchOrdersAsync(
```
Let's remove this method. It's not currently supported by the Coordinator server implementation.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+    }
+
+    /**
+     * Hard cancellation on Exchange contract. Cancels a single order.
```
```suggestion
     * Hard cancellation cancels an order on-chain and involves an Ethereum transaction.
```

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+    }
+
+    /**
+     * Hard cancellation on Exchange contract.
```
```suggestion
     * Batch hard cancellation on Exchange contract.
```

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        // );
+    }
+
+    public async assertValidCoordinatorApprovalsAsync(
```
If this is a public method, it needs doc comments.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        );
+    }
+
+    public async getSignerAddressAsync(hash: string, signature: string): Promise<string> {
```
Ditto re: doc comments.

fabioberger commented on packages/contract-wrappers/test/coordinator_wrapper_test.ts:
```diff
@@ -0,0 +1,359 @@
+
+        await app.listen(coordinatorServerConfigs.HTTP_PORT, () => {
+            // tslint:disable-line:await-promise
+            console.log(`Coordinator SERVER API (HTTP) listening on port ${coordinatorServerConfigs.HTTP_PORT}!`); // tslint:disable-line:no-console
```
Let's use the `logUtils.log` method here instead of `console.log`.

fabioberger commented on packages/contract-wrappers/test/coordinator_wrapper_test.ts:
```diff
@@ -0,0 +1,359 @@
+        );
+
+        await app.listen(coordinatorServerConfigs.HTTP_PORT, () => {
+            // tslint:disable-line:await-promise
```
This doesn't seem necessary.

fabioberger commented on packages/contract-wrappers/test/coordinator_wrapper_test.ts:
```diff
@@ -0,0 +1,359 @@
+            CoordinatorRegistry.compilerOutput.abi,
+            contractAddresses.coordinatorRegistry,
+            provider,
+            { gas: 6000000 },
```
Is this really needed?

xianny replied:
oops, that was leftover from when I was using the ganache snapshot in docker  instead of in-process ganache. removing.

fabioberger commented on packages/contract-wrappers/test/coordinator_wrapper_test.ts:
```diff
@@ -0,0 +1,359 @@
+            expect(result.status).to.be.equal(200);
+            expect(await result.text()).to.be.equal('pong');
+        });
+    });
```
Let's add some failure cases. Some that come to mind are:

- `feeRecipientAddress` is not in registry.
- `coordinatorEndpoint` in registry is incorrect. 

An additional edge-case:
- Batch fill with 2 orders each specifying a different coordinator server

xianny commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        data: string,
+        takerAddress: string,
+        feeRecipientAddress: string,
+        orderTransactionOpts: OrderTransactionOpts = { shouldValidate: true },
```
I don't have a strong opinion on this, but in every other place where it's accepted as an argument, this default is provided.

I'll defer to you on it if you still think we should remove it?

dekz replied:
I think it is fine to default in the `public` method. 

My take on having it defaulted in a `private` method is that a future developer could make a change and forget to pass the option from the user down into the `private` method resulting in unwanted behaviour. E.g it is specified as `false` in `hardCancelOrdersUpToAsync(...., false)` but we forget to pass that on and then becomes `true` in `_handleFillAsync` from the default. My thought is this makes it safer.

xianny replied:
makes sense to me!

xianny commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,672 @@
+        assert.isETHAddressHex('feeRecipientAddress', feeRecipientAddress);
+
+        const coordinatorOperatorEndpoint = await this._registryInstance.getCoordinatorEndpoint.callAsync(
+            feeRecipientAddress,
```
reimplementing

xianny commented:
I'd still like to write more tests for error cases to check the format of the errors, but apart from that I've implemented all the changes brought up. 

I changed everything to return `Promise<txhash | CoordinatorError>` at first, then changed it back to throw errors. But had to stringify `CoordinatorError` objects. Feeling ambivalent about this. If we return instead of throwing, it breaks the resolve/reject interface for Promises. Is there a way to throw an error with an object instead of just a string?

xianny commented:
tried to rebase on development and something funky happened. Managed to reset it but that's why there are weird events in the log.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+            if (serverEndpointsToFeeRecipients[endpoint] === undefined) {
+                serverEndpointsToFeeRecipients[endpoint] = [];
+            }
+            serverEndpointsToFeeRecipients[endpoint] = serverEndpointsToFeeRecipients[endpoint].concat(feeRecipient);
```
`push`

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+        for (const endpoint of Object.keys(serverEndpointsToFeeRecipients)) {
+            const response = await this._executeServerRequestAsync(transaction, takerAddress, endpoint);
+            if (response.isError) {
+                errorResponses = errorResponses.concat(response);
```
`push` everywhere.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+                });
+
+            // submit transaction with approvals
+            return this._submitCoordinatorTransactionAsync(
```
Let's assign and return so that it's clearer what is getting returned.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+            const approvedOrders = approvalResponses
+                .map(resp => {
+                    const endpoint = resp.coordinatorOperator;
+                    const feeRecipients: string[] = serverEndpointsToFeeRecipients[endpoint];
```
Is this type declaration `string[]` necessary?

xianny replied:
nope, removing

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+            // return errors and approvals
+            throw new CoordinatorServerError(
+                CoordinatorServerErrorMsg.CancellationFailed,
+                [],
```
For readability, our convention is to declare a variable with a descriptive name rather then passing in values directly. Which is more readable to you:

```
const orders = getOrders(2, [], false);
```

or

```
const limit = 2;
const excludedFields = [];
const isFilled = false;
const orders = getOrders(limit, excludedFields, isFilled);
```

?

xianny replied:
makes sense! will do

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+    }
+
+    /**
+     * Cancels an order on-chain and involves an Ethereum transaction.
```
```suggestion
     * Cancels an order on-chain by submitting an Ethereum transaction.
```

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+            order.makerAddress,
+            transaction.signature,
+            [],
+            [],
```
ditto above re: passing named variables.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+    }
+
+    /**
+     * Batch version of hardCancelOrderAsync. Cancels orders on-chain and involves an Ethereum transaction.
```
```suggestion
     * Batch version of hardCancelOrderAsync. Cancels orders on-chain by submitting an Ethereum transaction.
```

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+    }
+
+    /**
+     * Cancels orders on-chain and involves an Ethereum transaction.
```
```suggestion
     * Cancels orders on-chain by submitting an Ethereum transaction.
```

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+        await assert.isSenderAddressAsync('takerAddress', takerAddress, this._web3Wrapper);
+
+        const data = this._transactionEncoder.fillOrderTx(signedOrder, takerAssetFillAmount);
+        return this._handleFillsAsync(data, takerAddress, [signedOrder], orderTransactionOpts);
```
Assign then return.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+                serverEndpointsToFeeRecipients[endpoint] = [];
+            }
+            serverEndpointsToFeeRecipients[endpoint].push(feeRecipient);
+        }
```
This logic is duplicated in at least two locations. Let's functionally decompose and make it DRY.

xianny replied:
I simplified this part but left the rest separate (handling server responses) because slightly different things need to happen for fills vs cancels.

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+        try {
+            json = await response.json();
+        } catch (e) {
+            // ignore
```
Hm, this feel dangerous to me. When would this fail?

xianny replied:
this would fail if the server doesn't return a json response. e.g. something unexpected happens on server and it returns 500. I've had this happen a few times in testing, usually an uncaught exception from typeorm. 

This method basically tries to wrap server responses in a predictable format. The `json` field is either useful details or defaults to undefined, but `isError` and `status` are always informative.

fabioberger replied:
If it only happens when a non-200 is returned, I'd rather we do:

```
const json = isError ? undefined : await response.json();
```

That way, if we are expecting a body, and `.json()` throws, it won't be swallowed. Generally very cautious of swallowing thrown errors since a lot of bugs go undiscovered in this way.

xianny replied:
we actually expect a body for 200 and 400 but that's easy to express. Changing it!

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+    }
+
+    private async _submitCoordinatorTransactionAsync(
+        transaction: { salt: BigNumber; signerAddress: string; data: string },
```
Let's not in-line type declarations. This might by one of the reasons why your CI doc generation test is failing.

fabioberger commented on packages/contract-wrappers/src/types.ts:
```diff
@@ -225,3 +225,61 @@ export interface DutchAuctionData {
+            approvalSignatures: string[];
+            expirationTimeSeconds: BigNumber;
+            takerAssetFillAmount: BigNumber;
+        }
```
Make this a separate type declaration. Our docs cannot handle infinitely nested type definitions yet

fabioberger commented on packages/contract-wrappers/src/types.ts:
```diff
@@ -225,3 +225,61 @@ export interface DutchAuctionData {
+    error?: any;
+    request: {
+        signedTransaction: SignedZeroExTransaction;
+        txOrigin: string;
```
Ditto above here.

fabioberger commented on packages/contract-wrappers/src/types.ts:
```diff
@@ -225,3 +225,61 @@ export interface DutchAuctionData {
+    orders?: Array<SignedOrder | Order>;
+}
+
+export class CoordinatorServerError extends Error {
```
I wouldn't consider classes as types (as understood in Typescript). Let's move this to it's own file.

fabioberger commented on packages/contract-wrappers/src/types.ts:
```diff
@@ -225,3 +225,61 @@ export interface DutchAuctionData {
+    isError: boolean;
+    status: number;
+    body?: CoordinatorServerCancellationResponse | CoordinatorServerApprovalRawResponse;
+    error?: any;
```
Why `any`?

xianny replied:
Hmm. on second thought, this could be `ValidationError[]` with the type imported from the server. I guess the server will never return any json that isn't a success response known error type.

xianny replied:
Looking closer at the server response types, looks like the error response isn't exactly the formally defined type either. 

Response: https://github.com/0xProject/0x-coordinator-server/blob/master/ts/src/middleware/error_handling.ts#L34-L47

Exported type: https://github.com/0xProject/0x-coordinator-server/blob/master/ts/src/errors.ts#L19

Leaning towards leaving it as `any`, at least for now. This error field is never surfaced on its own, it's just for additional details from the server. The consumer always has access to other informative fields.

The data in this field will either be undefined or some kind of validation error so would have to be surfaced to end user anyway right? As in, even if we strongly type this field, it's not likely that the developer would be able to do automated error correction?

fabioberger commented on packages/contract-wrappers/test/coordinator_wrapper_test.ts:
```diff
@@ -0,0 +1,689 @@
+            // coordinator-server, as currently implemented, shares a singleton database connection across
+            // all instantiations. Making the request to two different mocked server endpoints still hits the
+            // same database and fails because of the uniqueness constraint on transactions in the database.
+            it.skip('should fill a batch of orders with different feeRecipientAddresses with different coordinator servers', async () => {
```
Will this still get fixed before we merge?

xianny replied:
No, this involves fixing the server side of things. I still think it's out of scope like we discussed. But will make a task to fix it after merging.

xianny commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+/**
+ * This class includes all the functionality related to filling or cancelling orders through
+ * the 0x V2 Coordinator smart contract.
+ * All logic related to filling orders is implemented in the private method `_handleFillsAsync`.
```
I'll remove it :)

xianny commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+            feeRecipientsToOrders[feeRecipient] = feeRecipientsToOrders[feeRecipient].concat(order);
+        }
+
+        const serverEndpointsToFeeRecipients: { [feeRecipient: string]: string[] } = {};
```
it's `endpoint -> feeRecipients`, the typedef was wrong. Fixing!

xianny commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,857 @@
+            feeRecipientAddress,
+        );
+        if (coordinatorOperatorEndpoint === '') {
+            throw new Error(
```
If `senderAddress = coordinator address` then it *has* to have a valid fee recipient address right? where "valid" means registered in the coordinator registry. That's also how Coordinator.sol works. 

I'll add a check to filter for orders that only have coordinator as the sender address, and throw if these coordinator orders don't have a valid fee recipient?

fabioberger replied:
Yep, I like that idea. Then users can batch fill OO and Coordinator orders.

xianny commented on packages/order-utils/src/index.ts:
```diff
@@ -68,7 +68,7 @@ export {
 
 export {
-    OrderError,
+    TypedDataError as OrderError,
```
@fabioberger should I export this as `TypedData` and add it to CHANGELOG?

fabioberger replied:
Yes, let's make this change.

fabioberger commented on packages/order-utils/src/signature_utils.ts:
```diff
@@ -278,7 +286,50 @@ export const signatureUtils = {
+        }
+    },
+    /**
+     * Signs an order using `eth_signTypedData` and returns a SignedOrder.
```
Update this comment to `transaction`

xianny replied:
ooops. thanks for catching

fabioberger commented on packages/order-utils/src/signature_utils.ts:
```diff
@@ -278,7 +286,50 @@ export const signatureUtils = {
+     *          must be available via the supplied Provider.
+     * @return  A SignedOrder containing the order and Elliptic curve signature with Signature Type.
+     */
+    async ecSignTypedDataTransactionAsync(
```
Let's also add a CHANGELOG entry for `order-utils` with a minor version bump detailing the addition of this method.

xianny replied:
shouldn't it be major version bump because we also change `OrderError` to `TypedDataError`?

fabioberger commented on packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts:
```diff
@@ -0,0 +1,855 @@
+ * This class includes all the functionality related to filling or cancelling orders through
+ * the 0x V2 Coordinator extension contract.
+ */
+export class CoordinatorWrapper extends ContractWrapper {
```
Add a CHANGELOG entry to `contract-wrappers` detailing the addition of this wrapper.

xianny commented on packages/contract-wrappers/src/utils/coordinator_server_types.ts:
```diff
@@ -0,0 +1,61 @@
+export interface CoordinatorServerApprovalRawResponse {
+    signatures: string[];
+    expirationTimeSeconds: BigNumber;
+}
```
the Coordinator contract expects signatures and expiration times to be a same length array. It iterates through signatures and compare to expiration times at the same index to decide whether the signature is still valid or not.

fabioberger replied:
Ah yep, makes sense.

xianny commented:
`static-tests-python` is failing everywhere else too. I'm going to call it unrelated and merge this. 🎉
